### PR TITLE
Add suport for running the heightgraph in embed mode

### PR DESCRIPTION
**Pull request by alexcojocaru**
* add additional option to disable the expand controls
* add support for resetting and redrawing the chart on data change,
  instead of doing a full reinitialization
* rename some element ids to make them more heightgraph specific,
  so that they don't conflict with potential other elements in the DOM
  with the same id
* add support for triggering the highlight of an item on the chart

### CODE REVIEW

File: src/L.Control.Heightgraph.js [hunk 1]
```diff
@@ -63,10 +66,18 @@ L.Control.Heightgraph = L.Control.extend({
...
-        this._init_options();
         this._prepareData();
         this._computeStats();
+
+        this._onAddData();
+    },
+    /**
+     * Trigger a re-render of the chart based on the existing data (e.g. on container resize).
+     */
+    _onAddData() {
```
Comment (alexcojocaru) [inline src/L.Control.Heightgraph.js:79]: I moved these calls out of `addData`, so that all this functionality can be called from the outside if the chart is to be re-rendered (but there is not need for full reinitialization).

File: src/L.Control.Heightgraph.js [hunk 2]
```diff
@@ -152,7 +169,7 @@ L.Control.Heightgraph = L.Control.extend({
      */
     _calculateFullExtent(data) {
         if (!data || data.length < 1) {
-            throw new Error("no data in parameters");
+            return null;
```
Comment (alexcojocaru) [inline src/L.Control.Heightgraph.js:172]: When running in embed mode and always expanded, the data array could be empty. Instead of erroring out, the heightgraph would still work by rendering nothing.

File: src/L.Control.Heightgraph.js [hunk 3]
```diff
@@ -461,25 +484,25 @@ L.Control.Heightgraph = L.Control.extend({
         this._focusDistance = this._focus.append("text")
             .attr("x", 7)
             .attr("y", -this._y(boxPosition) + textDistance)
-            .attr("id", "distance")
+            .attr("id", "heightgraph.distance")
```
Comment (alexcojocaru) [inline src/L.Control.Heightgraph.js:487]: In embed mode, there could be other elements on the page with these ids `distance`, `height`, etc. I gave them more specific ids to avoid conflicts.

File: src/L.Control.Heightgraph.js [hunk 4]
```diff
@@ -907,13 +932,58 @@ L.Control.Heightgraph = L.Control.extend({
                 this[param].style('display', 'none');
             }
     },
+    /*
+     * Handles the mouseover the map and displays distance and altitude level.
+     * Since this does a lookup of the point on the graph
+     * the closest to the given latlng on the provided event, it could be slow.
+     */
+    _mapMousemoveHandler(evt) {
```
Comment (alexcojocaru) [inline src/L.Control.Heightgraph.js:940]: This function is also for running in embed mode, when the outer wants to highlight a specific point on the chart on a given event (e.g. mouse over a point on the route).

File: src/L.Control.Heightgraph.js [hunk 5]
```diff
@@ -123,6 +136,10 @@ L.Control.Heightgraph = L.Control.extend({
             this._dragRectangleG.remove();
             this._dragRectangleG = null;
             this._dragRectangle = null;
+
+            // Performance improvement: we could cache the full extend when addData() is called
+            let fullExtent = this._calculateFullExtent(this._areasFlattended);
+            fullExtent && this._map.fitBounds(fullExtent);
```
Comment (alexcojocaru) [inline src/L.Control.Heightgraph.js:142]: Ask the map to zoom out so that the full route is visible.

Comment (boldtrn) [reply src/L.Control.Heightgraph.js:142]: Could explain what this change does?

Comment (boldtrn) [reply src/L.Control.Heightgraph.js:142]: BTW: yes I think caching would not hurt :+1:

File: src/L.Control.Heightgraph.js (hunk 3 above)
Comment (boldtrn) [reply src/L.Control.Heightgraph.js:487]: Excellent! Related to #65

File: src/L.Control.Heightgraph.js [hunk 6]
```diff
@@ -907,13 +932,58 @@ L.Control.Heightgraph = L.Control.extend({
...
+        let closestDistance = 2 * Math.pow(100, 2);
+
+        for (i = 0; i < this._areasFlattended.length; i++) {
+            let item = this._areasFlattended[i];
+
+            let latDiff = evt.latlng.lat - item.latlng.lat;
+            let lngDiff = evt.latlng.lng - item.latlng.lng;
+
+            // first check for an exact match; it's simple and avoid further calculations
+            if (latDiff == 0 && lngDiff == 0) {
```
Comment (boldtrn) [inline src/L.Control.Heightgraph.js:957]: Maybe we better check for diff <= .000001 or something like this? Exact matches with floating points numbers are guess work. In this case we might need to Math.abs(diff) before?

File: src/L.Control.Heightgraph.js (hunk 4 above)
Comment (boldtrn) [reply src/L.Control.Heightgraph.js:940]: I am just wondering, is this function used? And if yes, where is it used?

**Comment by boldtrn** [comment]
Thanks a lot! Could you have a look at my comments, I will see to give this PR a test.

So this PR is related to:
- #37 
- #65

**Comment by boldtrn** [comment]
Thanks for this PR, this seems to work fine.

File: src/L.Control.Heightgraph.js (hunk 4 above)
Comment (alexcojocaru) [reply src/L.Control.Heightgraph.js:940]: It's not used by the heightgraph code; it is a way for an external component to tell the heightgraph to highlight a point on the chart.

If your example at https://giscience.github.io/Leaflet.Heightgraph/, if I move the mouse cursor over the route on the map, nothing changes in the heightgraph.

With this function here, the map could handle the mouse over event (e.g. highlight the point on the route) and then call the heightgraph to highlight the corresponding point on the chart.

Let me know if this is still not clear, I could include an example of that.

File: src/L.Control.Heightgraph.js (hunk 5 above)
Comment (alexcojocaru) [reply src/L.Control.Heightgraph.js:142]: See https://mariusschulz.com/blog/the-and-and-or-operators-in-javascript#using-logical-operators-with-non-boolean-values

In this case, it executes the statement if and only if fullExtent is defined and not null (or any other value which would cast to a boolean false).

File: src/L.Control.Heightgraph.js (hunk 6 above)
Comment (alexcojocaru) [reply src/L.Control.Heightgraph.js]: done; let me know if it's better now

**Comment by TheGreatRefrigerator** [comment]
Hey @alexcojocaru ,

thanks a lot for the contribution!
From what i've seen, this looks solid.
Didn't have the chance to test it yet, as i am currently working on making this whole thing more development friendly and future proof as well ... (see rollup branch)

I would like to take a closer look at this pull in the next days.
The only thing i would suggest upfront is either use `event` or the abbreviation `e` for an event.
I think that's more common in js culture i think, but not quite sure...

best regards

File: src/L.Control.Heightgraph.js (hunk 5 above)
Comment (boldtrn) [reply src/L.Control.Heightgraph.js:142]: I can read and write js :). Why is the change added? What problem does it solve?

File: src/L.Control.Heightgraph.js (hunk 4 above)
Comment (boldtrn) [reply src/L.Control.Heightgraph.js:940]: Ah yes I thought so, it would be great if you could include it in the example and maybe add it to the  README as well?

**Comment by TheGreatRefrigerator** [comment]
Hi again @alexcojocaru ,
I have added some examples to test the addData() function with different data sets.
https://github.com/GIScience/Leaflet.Heightgraph/tree/easier_to_embed
Could you check this out and open the index-dev.html file in your browser:
One major problem with current structure is, if the data is updated, the mappings need to be updated as well.
So we would need additional functions for that.

You might want to wait on the rollup branch to get merged though, or base the changes directly on that one

File: src/L.Control.Heightgraph.js (hunk 5 above)
Comment (alexcojocaru) [reply src/L.Control.Heightgraph.js:142]: Oh, my bad then.

On selecting a subsection of the graph, the map zooms in or out to show only that portion of the route. On resetting the selection (aka the drag), the map should zoom out to show the full route.

Comment (alexcojocaru) [reply src/L.Control.Heightgraph.js:142]: This is the equivalent of `fitBounds()` call to the map in the `_fitSection` function below.

File: src/L.Control.Heightgraph.js (hunk 4 above)
Comment (alexcojocaru) [reply src/L.Control.Heightgraph.js]: Where in the README do you suggest I add that?

**Comment by alexcojocaru** [comment]
I am a bit unclear about this new branch you've created; it looks like I have to merge `master` into it, which is okay with me, and fix the issue around updating the mappings which you pointed out. Does that sound good to you?

On a related note: it looks like the `rollup` branch was already integrated; could you confirm that?

**Comment by TheGreatRefrigerator** [comment]
> Does that sound good to you?
Having to update both mappings and data seperately still doesn't sound really ideal from a workflow point of view for me.
But i also can't think of a better way right now.
So feel free to go ahead and implement it that way 👍 

> let me know how I should go about resolving them
`index-dev.html` -> moved to: `/example/index.html` **can be removed**
logic part of `index.html` -> moved to `/example/example.js` (just to avoid having it in both index files as they currently do not differ) **if you didn't change index.html accept from master**
**package-lock.json should be used from the master branch, as it removes a lot of unnecessary dependencies.**

If you need help, i will merge it.

**Comment by alexcojocaru** [comment]
Finished merging master into my feature branch. I started the simple server with `npm start` and the dataset renders correctly.


You said this in a previous comment:

> I have added some examples to test the addData() function with different data sets.

hmm... I see only one at http://localhost:8080/example

> https://github.com/GIScience/Leaflet.Heightgraph/tree/easier_to_embed
> Could you check this out and open the index-dev.html file in your browser:

Do I just open the html file explicitly in the browser? isn't this the file that gets rendered at http://localhost:8080/example?

> One major problem with current structure is, if the data is updated, the mappings need to be updated as well.
> So we would need additional functions for that.

Could you describe a test case for that situation? I have not tested the latest version explicitly, but I did test the one in the original pull request thoroughly and the height chart rendered correctly on data changes. For testing all this, I would need a relatively elaborate setup, and so I am trying to sort out all the merge and integration issues first, before looking into testing that.
More specifically, I am trying to get the height chart integrated into the http://brouter.de/brouter-web/

**Comment by TheGreatRefrigerator** [comment]
> Do I just open the html file explicitly in the browser? isn't this the file that gets rendered at http://localhost:8080/example?

Ah no, i did add the examples for this usecase only to the easier_to_embed branch in Leaflet.Heightgraph. See commit:
1b1af00963862fd569afc1ea0c4693fd6a866325

Should have done that in your repo instead.
If you look at this ^ commit and integrate the changes, you should have the examples you need.

It basically adds this block https://github.com/GIScience/Leaflet.Heightgraph/blob/easier_to_embed/index-dev.html#L27-L46
And adds the additional example files.
If you add that to /example/index.html & example/example.js. and run `npm start` you should be able to test this.

**Comment by alexcojocaru** [comment]
Got my local env working and tested the three sets of sample data.

The first one looks good.
The second one is broken due to the wrong `summary` (the value should be `steepness`, all lowercase). Once I fixed that, it rendered correctly.
The third one looks good too - the chart is empty.
The one issue I have noticed is that,  when switching from dataset 1 to dataset 2, the chart type doesn't switch to steepness - I was expecting that, for dataset 2 only has the steepness feature collection and not surface.

Given the above, I am not following this comment of yours. Could you provide more info?
> One major problem with current structure is, if the data is updated, the mappings need to be updated as well.
So we would need additional functions for that.

**Comment by TheGreatRefrigerator** [comment]
> The one issue I have noticed is that, when switching from dataset 1 to dataset 2, the chart type doesn't switch to steepness

Yes, i see what you mean.
That should be addressed in this change.

I didn't find a branch with the integrated examples on the latest development in your fork, so i just rebased this and tweaked the examples a bit:
https://github.com/GIScience/Leaflet.Heightgraph/tree/easier_to_embed_rebased
Please use this branch for further development on this.

I have added another dataset to explain the problem (checkout the above branch and run `npm start`, you should see 4 examples now):

The first data set has 4 categories (surface, steepness, waytype, suitability). All of them were reflected in the mappings passed to the heightgraph initially.

the second data set has only 1 category (steepness) which is included in the original set of mappings. That's why it's working without a problem.

Data set 3 has 1 category (conditionally_closed) which is **not** included in the original mappings for the first dataset.
There would be an error if the mappings for this dataset are not added initially (i added them already, so the 3rd example is working already: https://github.com/GIScience/Leaflet.Heightgraph/commit/111a4a18972eb9ae86d51615fbd42d8f3d425ed6#diff-c552d91cc708fa1d9710065a30f2e5cfR205-R214)

But if we go for adding Data dynamically, after the heightgraph is initialized, i think we should do the same for the mappings. So when the data set changes, the mappings should change too, to only contain mappings for the current data set. Otherwise we could also add all datasets during initialization and just make it switch between them.

Does this make sense to you?

**Comment by alexcojocaru** [comment]
Thanks for setting up the rebased branch, I've been working on it.
The good news is that I am getting rid of the _onAddData() helper (https://github.com/GIScience/Leaflet.Heightgraph/blob/easier_to_embed_rebased/src/L.Control.Heightgraph.js#L117), as you've added a dedicated function to resize the chart, hence the mappings issue is fixed.
There is another issue with existing selections on the chart on data change, which is fixed locally.
The last known issue is the category name / selection box, which I am working on fixing.
Once the above is complete, I will resurrect this PR. I would prefer that the PR is to merge into your `easier_to_embed_rebased` branch, and let you merge it into master after the PR is integrated. Does this work for you?

**Comment by TheGreatRefrigerator** [comment]
> Once the above is complete, I will resurrect this PR. I would prefer that the PR is to merge into your `easier_to_embed_rebased` branch, and let you merge it into master after the PR is integrated. Does this work for you?

Yes that would be great!

> as you've added a dedicated function to resize the chart, hence the mappings issue is fixed.

Is the the zooming to full extent, when clearing the selection (with single click on the heightgraph after drag selecting a part) still working then?

Thanks for helping with this!

File: src/L.Control.Heightgraph.js [hunk 7]
```diff
@@ -106,20 +106,17 @@ import {
                     .remove();
             }
 
+            this._removeMarkedSegmentsOnMap();
+            this._resetDrag(false);
+
             this._data = data;
+            this._init_options();
```
Comment (alexcojocaru) [inline src/L.Control.Heightgraph.js:113]: I removed the helper (which I introduced in a previous commit) and added this call to `_init_options()` back where it was. No more mapping issues.

File: src/L.Control.Heightgraph.js [hunk 8]
```diff
@@ -106,20 +106,17 @@ import {
...
-        /**
-         * Trigger a re-render of the chart based on the existing data (e.g. on container resize).
-         */
-        _onAddData() {
-            this._init_options();
             this._appendScales();
             this._appendGrid();
             this._createChart(this._selectedOption);
-            if (this._data.length > 1) this._createSelectionBox();
+            this._createSelectionBox();
```
Comment (alexcojocaru) [inline src/L.Control.Heightgraph.js:119]: this function is self contained now; it can deal with empty data set

File: src/L.Control.Heightgraph.js [hunk 9]
```diff
@@ -183,17 +180,21 @@ import {
             }
         },
         /**
-         * Removes the drag rectangle and zooms back to the total extent of the data.
+         * Removes the drag rectangle
+         * @param {boolean} skipMapFitBounds - whether to zoom the map back to the total extent of the data 
          */
-        _resetDrag() {
+        _resetDrag(skipMapFitBounds) {
```
Comment (alexcojocaru) [inline src/L.Control.Heightgraph.js:186]: I introduced the arg as is in order to keep the function backwards compatible; if the function is called with no arg (i.e. all existing calls to it), it's as if the caller want to fit the map to bounds (i.e. the previous behavior). The caller would have to explicitly pass `true` as arg to skip the fitBounds call from being made.

**Comment by alexcojocaru** [comment]
> > Once the above is complete, I will resurrect this PR. I would prefer that the PR is to merge into your `easier_to_embed_rebased` branch, and let you merge it into master after the PR is integrated. Does this work for you?
> 
> Yes that would be great!

Done

> 
> > as you've added a dedicated function to resize the chart, hence the mappings issue is fixed.
> 
> Is the the zooming to full extent, when clearing the selection (with single click on the heightgraph after drag selecting a part) still working then?
> 
> Thanks for helping with this!

And done

I've added more comments on this PR to help explain some of the changes  I made.

If there is something else missing, let me know. One thing which still stands out from the review on the initial commit is boldtrn's request to document the `_mapMousemoveHandler()` function. Since there is no doc in the README on the publicly exposed functions, I have not done that, but instead I added doc on the code: https://github.com/alexcojocaru/Leaflet.Heightgraph/blob/easier_to_embed/src/L.Control.Heightgraph.js#L996

**Comment by TheGreatRefrigerator** [comment]
I encountered 2 more issues both to do with zooming out after selecting a segment:

- it zoomes out every time when i change the category. In this case the selection should be kept.
- when calling hg.addData(dataSet), it zooms out to the last dataset instead of the new one.

I'm still not quite sure how the `_mapMousemoveHandler()` is working exactly. Is it possible to add this to the existing example code?
If it's called from outside, i think it shouldn't have the underscore as it indicates a private methods.

**Comment by alexcojocaru** [comment]
Re issue #1 in your comment:
There is a similar problem in the master branch: the selection is removed, but the map is not zoomed out. If anything, the version in this PR zooms the map out, which is expected when the selection is removed.
It appears that this behavior is due to the mouse click event bubbling up and triggering a selection reset.

**Comment by alexcojocaru** [comment]
Re the 2nd issue in your last comment: it's fixed now, thanks for catching it.

**Comment by alexcojocaru** [comment]
> I'm still not quite sure how the `_mapMousemoveHandler()` is working exactly. Is it possible to add this to the existing example code?
> If it's called from outside, i think it shouldn't have the underscore as it indicates a private methods.

I've renamed it and added an example of using it; move the mouse cursor over the route and see the corresponding point on the chart being highlighted.

**Comment by alexcojocaru** [comment]
Thanks for your help with getting this integrated, especially with testing all the edge cases.

**Comment by TheGreatRefrigerator** [comment]
No problem,
Thank you for adding this!

I don't think i tested _all_ the edge cases but i didn't find anything else for now ;)
In the long run i would like to add tests for everything + example test templates so people can also add tests to new features. But everything needs time -.-

It's a lot more fun to work together on code.
